src/hbase/test-util/mini-cluster.cc

```diff
@@ -93,7 +97,8 @@ MiniCluster::~MiniCluster() {
 }
 
 void MiniCluster::Setup() {
```

Comment:
I think MiniCluster's usage should be declarative. If it is instantiated it will be used. As it stands, the JVM creation and assignment is performed lazily, but as you pointed out, it creates a state management issue. If we declare a MiniCluster I think we should create the JVM and class/method objects in the constructor. This will allow the destructor to destroy the java MiniCluster. Is there a case where MiniCluster is created without the Java based mini cluster is used? 

I may be missing something..but it seems that the code would be dramatically simplified if Setup is always performed in the constructor and StopCluster is always performed in the destructor. Then the caller would be responsible for calling methods that would start/stop the cluster.

Reply:
I think this ties to my response to the other comment. My understanding is that the cluster lifecycle could potentially be a little more complex than just start/stop, hence the mutex. Let me know if you think its an overkill and I can redo the patch. I agree with you that it simplifies the overall patch.